libresolv: fix build on aarch64-darwin

reckenrode opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

`libresolv` doesn’t build on aarch64-darwin due to the the 11.0 SDK replacing `configd` with a version that doesn’t include the header needed to build `libresolv`. I’ve introduced an additional, header-only derivation (`configdHeaders`) that provides the headers while leaving the version from the SDK as it is. If it’s not the best approach, I’m open to other options.

This should fix #134292.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

reckenrode committed:
configdHeaders: init at 453.19

`configd` is aliased to `apple-sdk.frameworks.SystemConfiguration` in
apple-sdk-11.0, which is the default on aarch64-darwin, but it does not
include all of the headers.  This package introduces the missing
headers, which are needed to build `libresolv`.

A subsequent commit will fix libresolv to build on aarch64-darwin.

reckenrode committed:
libresolv: fix build on aarch64-darwin

Use `configdHeaders` instead of `configd`.  This fixes #134292.

reckenrode commented:
Result of `nixpkgs-review pr 147453` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>dnstracer</li>
    <li>gpg-tui</li>
    <li>onefetch</li>
    <li>weechat</li>
    <li>weechat-unwrapped</li>
  </ul>
</details>

reckenrode commented:
I’ve started `nixpkgs-review pr` for x86_64-darwin. It’s going to be a while.

veprbl commented:
Didn't work, might have to target staging

veprbl commented on pkgs/os-specific/darwin/apple-source-releases/default.nix:
```diff
@@ -238,7 +238,7 @@ let
 in
 
-developerToolsPackages_11_3_1 // macosPackages_11_0_1 // {
+developerToolsPackages_11_3_1 // macosPackages_11_0_1 // rec {
```
Relying on `rec` prevents you from being able to properly propagate overrides

reckenrode replied:
Okay. I can update it to call `applePackage` like the other derivations. It shouldn’t use `pkgs.darwin.configd` because that will get replaced by the SDK package on `aarch64-darwin`, which is the problem that’s being worked around. Should I add a comment explaining why it’s not following the pattern of the other headers-only variants?

Also, would it be better to default `Security` to `null` in the call or pass it in even though it’s not going to be used in the headers-only derivation?

SuperSandro2000 commented on pkgs/os-specific/darwin/apple-source-releases/configd/default.nix:
```diff
@@ -63,6 +64,8 @@ appleDerivation' stdenv {
     EOF
   '';
 
+  ${if headersOnly then "dontBuild" else null} = headersOnly;
```
You are sure this evals when headersOnly is false?

veprbl replied:
It definitely evals, although I never quiet understood how/if the trick works:
```
nix-repl> with import <nixpkgs> {}; stdenv.mkDerivation { name = "test"; }         
«derivation /nix/store/25nj2qsxjzrbnwjp924nfr0bfjksjf40-test.drv»

nix-repl> with import <nixpkgs> {}; stdenv.mkDerivation { name = "test"; null = 1; }
«derivation /nix/store/7zylvkwsk0gszax0x3i16ybfwsh3xrsr-test.drv»
```

reckenrode commented:
> Didn't work, might have to target staging

`libresolv` changed to take `configdHeaders` instead of `configd`, so I don’t think a rebuild is avoidable. I wonder whether [this conditional][1] should be made unconditional once `libresolv` is building on aarch64-darwin.

[1]: https://github.com/NixOS/nixpkgs/blob/94d91a448b87a70204485bd768977c07575911e8/pkgs/stdenv/darwin/make-bootstrap-tools.nix#L88

veprbl commented:
@reckenrode Could you explain a bit, what is the issue here? You've said that the `libresolv` doesn't build on aarch64-darwin, so I assumed that the we might be using an impure libresolv from /usr/lib instead of the Open Source build, but that's not the case. How can `libresolv` be broken if  thestdenv is not?

reckenrode commented:
> Could you explain a bit, what is the issue here?

A handful of packages depend explicitly on `libresolv` regardless of platform (see #134292). Like I said, I’m open to other solutions (including changing those derivations not to remove the explicit dependency).

> You've said that the `libresolv` doesn't build on aarch64-darwin, so I assumed that the we might be using an impure libresolv from /usr/lib instead of the Open Source build, but that's not the case. How can `libresolv` be broken if the stdenv is not?

As far as I can tell, `libresolv` on aarch64-darwin is the system one. The `installname` in the `tbd` file points to `/usr/lib/libresolv.9.dylib`. You can see the difference comparing binaries linked for the two platforms. I assume `stdenv` that’s why `stdenv` is not broken on aarch64-darwin.

```
# aarch64-darwin
$ otool -L (command -v rg)
/etc/profiles/per-user/reckenrode/bin/rg:
	/nix/store/4cq6zxqlgvi7sj07lmr0bv1y8rl2ycp1-pcre2-10.36/lib/libpcre2-8.0.dylib (compatibility version 11.0.0, current version 11.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/nix/store/c7lpgqfw4izdxspdy8s94rd9hwlpw78s-libiconv-50/lib/libiconv.dylib (compatibility version 7.0.0, current version 7.0.0)

# x86_64-darwin
/etc/profiles/per-user/reckenrode/bin/rg:
	/nix/store/6qa9a0rdraqzqj004xgk5wnnhkh65n4d-pcre2-10.36/lib/libpcre2-8.0.dylib (compatibility version 11.0.0, current version 11.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1238.60.2)
	/nix/store/afpdl5r9jnb8bx42nzpqi0c6g1flwldg-Libsystem-1238.60.2/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
	/nix/store/015rgxn5xx5awkz48yhh9c9n1bsvvivm-libiconv-50/lib/libiconv.dylib (compatibility version 7.0.0, current version 7.0.0)
```

veprbl commented:
@reckenrode That makes sense. It should be fine to put this on staging then. It would be nice to change the aarch64-darwin stdenv to use this fixed source build now.

domenkozar commented:
cc @toonn

toonn reviewed (commented):
LGTM though I can't test it on aarch64-darwin.

I do think a clarifying comment on the configd.override would be welcome.

reckenrode commented:
Result of `nixpkgs-review pr 147453` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)

reckenrode commented:
I updated the PR based on the feedback and added a comment explaining the `configdHeaders` override. `nixpkgs-review` is building on aarch64-darwin now.
